cpp/include/cuspatial/experimental/detail/linestring_distance.cuh

```diff
@@ -0,0 +1,284 @@
+
+  auto const p1_idx = threadIdx.x + blockIdx.x * blockDim.x;
+  size_t const num_linestrings =
+    thrust::distance(linestring1_offsets_begin, linestring1_offsets_end);
+
+  size_t const linestring1_num_points =
+    thrust::distance(linestring1_points_begin, linestring1_points_end);
+  size_t const linestring2_num_points =
+    thrust::distance(linestring2_points_begin, linestring2_points_end);
+
+  if (p1_idx >= linestring1_num_points) { return; }
+
+  size_t const linestring_idx =
+    thrust::distance(linestring1_offsets_begin,
+                     thrust::upper_bound(
+                       thrust::seq, linestring1_offsets_begin, linestring1_offsets_end, p1_idx)) -
+    1;
+
+  auto ls1_end = endpoint_index_of_linestring(
+    linestring_idx, linestring1_offsets_begin, num_linestrings, linestring1_num_points);
+
+  if (p1_idx == ls1_end) {
+    // Current point is the end point of the line string.
+    return;
+  }
+
+  auto ls2_start = *(linestring2_offsets_begin + linestring_idx);
+  auto ls2_end   = endpoint_index_of_linestring(
+    linestring_idx, linestring2_offsets_begin, num_linestrings, linestring2_num_points);
+
+  auto const& A = thrust::raw_reference_cast(linestring1_points_begin[p1_idx]);
+  auto const& B = thrust::raw_reference_cast(linestring1_points_begin[p1_idx + 1]);
+
+  auto min_squared_distance = std::numeric_limits<T>::max();
+  for (auto p2_idx = ls2_start; p2_idx < ls2_end; p2_idx++) {
+    auto const& C        = thrust::raw_reference_cast(linestring2_points_begin[p2_idx]);
+    auto const& D        = thrust::raw_reference_cast(linestring2_points_begin[p2_idx + 1]);
+    min_squared_distance = std::min(min_squared_distance, squared_segment_distance(A, B, C, D));
+  }
+
+  atomicMin(&thrust::raw_reference_cast(*(distances + linestring_idx)),
+            static_cast<T>(std::sqrt(min_squared_distance)));
+}
+
+}  // namespace detail
+
+template <class Cart2dItA, class Cart2dItB, class OffsetIterator, class OutputIt>
+void pairwise_linestring_distance(OffsetIterator linestring1_offsets_first,
+                                  OffsetIterator linestring1_offsets_last,
+                                  Cart2dItA linestring1_points_first,
+                                  Cart2dItA linestring1_points_last,
+                                  OffsetIterator linestring2_offsets_first,
+                                  Cart2dItB linestring2_points_first,
+                                  Cart2dItB linestring2_points_last,
+                                  OutputIt distances_first,
+                                  rmm::cuda_stream_view stream)
+{
+  using T = typename std::iterator_traits<Cart2dItA>::value_type::value_type;
+
+  static_assert(
+    detail::is_floating_point<T,
+                              typename std::iterator_traits<Cart2dItB>::value_type::value_type,
+                              typename std::iterator_traits<OutputIt>::value_type>(),
+    "Inputs and output must be floating point types.");
```

Comment:
I'm not sure of the best wording, but ideally this should clarify that it's referring to the underlying type of the `cartesian_2` to differentiate from the third error check below.

Reply:
```suggestion
    "Inputs and output must have floating point value type.");
```